src/passes/DeadStoreElimination.cpp

```diff
@@ -0,0 +1,678 @@
+        return aConst->value == bConst->value;
+      }
+    }
+    return false;
+  }
+};
+
+// Parent class of all implementations of the logic of identifying stores etc.
+// One implementation of Logic can handle globals, another memory, and another
+// GC, etc., implementing the various hooks appropriately.
+struct Logic {
+  Function* func;
+
+  Logic(Function* func, PassOptions& passOptions, FeatureSet features)
+    : func(func) {}
+
+  //============================================================================
+  // Hooks to identify relevant things to include in the analysis.
+  //============================================================================
+
+  // Returns whether an expression is a store.
+  //
+  // The main code will automatically ignore unreachable (irrelevant) stores.
+  bool isStore(Expression* curr) { WASM_UNREACHABLE("unimp"); };
+
+  // Returns whether an expression is a load.
+  //
+  // The main code will automatically ignore unreachable (irrelevant) loads.
+  bool isLoad(Expression* curr) { WASM_UNREACHABLE("unimp"); };
+
+  // Returns whether the expression is a barrier to our analysis: something that
+  // we should stop when we see it, because it could do things that we cannot
+  // analyze. A barrier will definitely pose a problem for us, as opposed to
+  // something mayInteract() returns true for - we will check for interactions
+  // later for mayInteract()s, but with barriers we don't need to.
+  //
+  // The default behavior here considers all calls to be barriers. Subclasses
+  // can use whole-program information to do better.
+  bool isBarrier(Expression* curr, const ShallowEffectAnalyzer& currEffects) {
+    // TODO: ignore throws of an exception that is definitely caught in this
+    //       function
+    // TODO: if we add an "ignore after trap mode" (to assume nothing happens
+    //       after a trap) then we could stop assuming any trap can lead to
+    //       access of global data, likely greatly reducing the numer of
+    //       barriers.
+    return currEffects.calls || currEffects.throws || currEffects.trap ||
+           currEffects.branchesOut;
+  };
+
+  // Returns whether an expression may interact with loads and stores in
+  // interesting ways. This is only called if isStore(), isLoad(), and
+  // isBarrier() all return false; that is, if we cannot identify the expression
+  // as one of those simple categories, this allows us to still care about it in
+  // our analysis.
+  bool mayInteract(Expression* curr, const ShallowEffectAnalyzer& currEffects) {
+    WASM_UNREACHABLE("unimp");
+  }
+
+  //============================================================================
+  // Hooks that run during the analysis
+  //============================================================================
+
+  // Returns whether an expression is a load that corresponds to a store, that
+  // is, that loads the exact data that the store writes.
+  bool isLoadFrom(Expression* curr,
```

Comment:
Is there a reason we distinguish `isLoadFrom` and `mayInteract(With)`? From dead store elimination's point, aren't they the same?